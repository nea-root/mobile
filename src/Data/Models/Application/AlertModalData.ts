export type AlertModalData = {
  isShow: boolean
  title?: string
  description?: string
  testID?: string
  actions?: AlertModalActionData[]
}

export type AlertModalActionData = {
  testID?: string
  label: string
  action: () => void
}
