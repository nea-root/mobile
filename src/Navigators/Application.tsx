import * as React from 'react';
import { View, Text, TouchableOpacity, Pressable } from 'react-native'
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { MasterContext } from '@/Context/MasterContext';
import { AuthStacks, RootStacks, Tabs, UserStacks } from './utils';

import Walkthrough from '@/Containers/Walkthrough/Walkthrough';
import OnBoardingScreen from '@/Containers/OnBoarding/OnBoardingScreen';
import { FlowContext } from '@/Context/Types/FlowContext';
import { FlowProvider } from '@/Context/FlowProvider/FlowProvider';
import Register from '@/Containers/Authentication/Register/Register';
import Header from './Header';
import Verification from '@/Containers/Verification/Verification';
import Login from '@/Containers/Authentication/Login/Login';
import { AuthProvider, useAuth } from '@/Context/AuthProvider/AuthProvider';
import { signOut } from '@/Services/Authentication/AuthService';

export type RootStackParamList = {
    [RootStacks.OnBoarding]: undefined;
    [RootStacks.Walkthrough]: { flowType: string };
};
export type UserStackParamList = {
    [UserStacks.Walkthrough]: undefined;
    [UserStacks.AuthStack]: { screen: keyof AuthStackParamList } | undefined;
};
export type AuthStackParamList = {
    [AuthStacks.Register]: undefined
    [AuthStacks.Login]: undefined
    [AuthStacks.Verification]: { formData: { username: string, email: string, password: string, role: string } }
}


const RootStack = createNativeStackNavigator<RootStackParamList>();
const Tab = createBottomTabNavigator();
const UserStack = createNativeStackNavigator<UserStackParamList>();
const AuthStack = createNativeStackNavigator<AuthStackParamList>();

// Dummy Screems
const DummyScreen = () => {
    const { logout } =  useAuth()
    const { flowType } = React.useContext(FlowProvider)
    return (
        <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
            <Pressable onPress={async ()=>{
                if(flowType){
                    await signOut(flowType)
                    logout(flowType)
                }
            }} style={{height: 40, width: '80%', borderColor: '#000', borderWidth: 1, alignContent: 'center', justifyContent: 'center' }}><Text style={{alignSelf: 'center', textAlign: 'center'}}>Logout</Text></Pressable>
        </View>
    );
}

const AuthStackNavigator = () => {
    return (
        <AuthStack.Navigator screenOptions={{ headerShown: false }}>
            <AuthStack.Screen name={AuthStacks.Register} component={Register} />
            <AuthStack.Screen name={AuthStacks.Login} component={Login} />
            <AuthStack.Screen name={AuthStacks.Verification} component={Verification} />
        </AuthStack.Navigator>)
}

const UserStackNavigator = ({ route }: any) => {
    const [isLoggedIn, setIsLoggedIn] = React.useState<boolean>(false)
    const { flowType, setFlowType }: FlowContext = React.useContext(FlowProvider)
    const { authState } = useAuth()
    React.useEffect(() => {
        setFlowType(route?.params?.flowType)
    }, [setFlowType])
    React.useEffect(() => {
        if (authState && authState.roles.length > 0 && flowType && authState.users[flowType]) {
            setIsLoggedIn(true)
        }
        else if(authState && authState.roles.length > 0 && flowType && !authState.users[flowType]){
            setIsLoggedIn(false)
        }
    }, [authState, authState.users, authState.tokens])
    if (isLoggedIn) {
        return (<Tab.Navigator screenOptions={{
            header: ({ navigation, route, options }) => (
                <Header
                    onLeftPress={navigation.goBack} // Handles back navigation
                    onRightPress={() => console.log("Right Action Pressed")} title={''} />
            ),
        }}>
            <Tab.Screen name={Tabs.Home} component={DummyScreen} />
            <Tab.Screen name={Tabs.Awareness} component={DummyScreen} />
            <Tab.Screen name={Tabs.Evidence} component={DummyScreen} />
            <Tab.Screen name={Tabs.Help} component={DummyScreen} />
        </Tab.Navigator>)
    }
    return (
        <UserStack.Navigator screenOptions={{
            header: ({ navigation, route, options }) => (
                <Header
                    onLeftPress={navigation.goBack} // Handles back navigation
                    onRightPress={() => console.log("Right Action Pressed")} title={''} />
            ),
            headerTransparent: true
        }}>
            <UserStack.Screen name={UserStacks.Walkthrough} component={Walkthrough} />
            <UserStack.Screen name={UserStacks.AuthStack} component={AuthStackNavigator} />
        </UserStack.Navigator>
    )
}



const ApplicationNavigator = () => {
    const { flowType, setFlowType } = React.useContext(FlowProvider)
    return (
        <NavigationContainer>
            <MasterContext>
                <RootStack.Navigator screenOptions={{ headerShown: false }}>
                    <RootStack.Screen name={RootStacks.OnBoarding} component={OnBoardingScreen} />
                    <RootStack.Screen name={RootStacks.Walkthrough} component={UserStackNavigator} />
                </RootStack.Navigator>
                {/* <Tab.Navigator>
                    <Tab.Screen name={Tabs.Home} component={DummyScreen} />
                    <Tab.Screen name={Tabs.Awareness} component={DummyScreen} />
                    <Tab.Screen name={Tabs.Evidence} component={DummyScreen} />
                    <Tab.Screen name={Tabs.Help} component={DummyScreen} />
                </Tab.Navigator> */}
            </MasterContext>
        </NavigationContainer>
    );
}

export default ApplicationNavigator;
