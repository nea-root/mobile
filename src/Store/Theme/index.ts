import { createSlice } from '@reduxjs/toolkit'
import themes from '@/Theme/themes'
import themeState from '@/Store/States/ThemeState'

export const initialState = themeState
const ThemeSlice = createSlice({
  name: 'theme',
  initialState: initialState,
  reducers: {
    changeTheme: (state, { payload: { theme, darkMode } }: ThemePayload) => {
      if (typeof theme !== 'undefined') {
        state.theme = theme
      }
      if (typeof darkMode !== 'undefined') {
        state.darkMode = darkMode
      }
    },
    setDefaultTheme: (
      state,
      { payload: { theme, darkMode } }: ThemePayload,
    ) => {
      if (!state.theme) {
        if (typeof theme !== 'undefined') {
          state.theme = theme
        }
        if (typeof darkMode !== 'undefined') {
          state.darkMode = darkMode
        }
      }
    },
  },
})

export const { changeTheme, setDefaultTheme } = ThemeSlice.actions

export default ThemeSlice.reducer

type DarkProps<T> = {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  [K in keyof T]: K extends `${infer Prefix}_dark` ? K : never
}[keyof T]

type PropsWithoutDark<T> = Omit<T, DarkProps<T>>

export type ThemeState = {
  theme: 'default' | keyof PropsWithoutDark<typeof themes>
  darkMode: boolean | null
}

type ThemePayload = {
  payload: Partial<ThemeState>
}
