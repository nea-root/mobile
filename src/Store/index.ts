
import AsyncStorage from '@react-native-async-storage/async-storage'
import { configureStore, combineReducers } from '@reduxjs/toolkit'
import { setupListeners } from '@reduxjs/toolkit/query'
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist'

import theme from './Theme'
import dummy from './Slices/DummySlice'
import { useDispatch, TypedUseSelectorHook, useSelector } from 'react-redux'


const reducers = combineReducers({
    theme,
    dummy
})

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
  whitelist: ['theme'],
}

const persistedReducer = persistReducer(persistConfig, reducers)

const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware => {
    const middlewares = getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    })

    // // @ts-ignore
    // if (__DEV__ && !process.env.JEST_WORKER_ID) {
    //   const createDebugger = require('redux-flipper').default
    //   middlewares.push(createDebugger())
    // }

    return middlewares
  },
})

const persistor = persistStore(store)

setupListeners(store.dispatch)

export type StoreStateType = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export const useAppDispatch: () => AppDispatch = useDispatch
export const useSelectorHook: TypedUseSelectorHook<StoreStateType> = useSelector

export { store, persistor }
