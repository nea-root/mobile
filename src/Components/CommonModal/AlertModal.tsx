import { AlertModalData } from '@/Data/Models/Application/AlertModalData'
import { useTheme } from '@/Hooks'
import { Modal, Text, View, ViewStyle } from 'react-native'
import { Button, Divider } from 'react-native-paper'
import createStyles from './AlertModalStyle'

export const AlertModal = (props: AlertModalData) => {
  const { isShow, title, description, actions } = props
  const theme = useTheme()
  const styles = createStyles(theme)

  return (
    <Modal testID="alertModal" visible={isShow} transparent={true}>
      <View style={styles.centeredView}>
        <View style={styles.modalView}>
          {title && title.length > 0 && (
            <Text style={[styles.titleText]}>{title}</Text>
          )}
          <Text style={[styles.descriptionText]}>{description}</Text>
          {actions && (
            <>
              <Divider style={[styles.divider]} />
              <View style={styles.actionContainer}>
                {actions?.map((data, index) => {
                  return (
                    <Button
                      key={`${data.label}_${index}`}
                      style={styles.actionButton}
                      labelStyle={styles.textStyle}
                      removeClippedSubviews={false}
                      testID={
                        'alert_action_button_' +
                        (data.testID ? data.testID : data.label)
                      }
                      onPress={data.action}
                    >
                      {data.label}
                    </Button>
                  )
                })}
              </View>
            </>
          )}
        </View>
      </View>
    </Modal>
  )
}

export const CONTAINER: ViewStyle = {}
