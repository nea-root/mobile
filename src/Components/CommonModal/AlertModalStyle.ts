import { CommonParams } from '@/Theme/theme'
import { StyleSheet } from 'react-native'

export default function <C>({
  Colors,
  Layout,
  Gutters,
  Fonts,
  MetricsSizes,
  ViewSizes,
  Dimens,
}: CommonParams<C>) {
  return StyleSheet.create({
    centeredView: {
      ...Layout.fill,
      ...Layout.center,
    },
    modalView: {
      ...Layout.alignSelfStretch,
      paddingTop: MetricsSizes.large,
      marginHorizontal: MetricsSizes.large,
      backgroundColor: Colors.Palette.white,
      borderRadius: Dimens.modalBorderRaidus,
      shadowColor: Colors.alertShadowColor,
      shadowOffset: {
        width: Dimens.zero,
        height: Dimens.alertShadowOffsetHeight,
      },
      shadowOpacity: Dimens.alertShadowOpacity,
      shadowRadius: Dimens.alertShadowRadius,
      elevation: Dimens.alertShadowElevation,
    },
    titleText: {
      textAlign: 'center',
      ...Fonts.textBoldRegular,
      color: Colors.alertTextColor,
      marginHorizontal: Dimens.alertMarginTop,
    },
    descriptionText: {
      ...Fonts.textSmallBold,
      color: Colors.alertTextColor,
      ...Gutters.largeHMargin,
      marginTop: Dimens.alertDescriptionMarginTop,
      textAlign: 'center',
    },
    divider: {
      ...Layout.alignSelfStretch,
      marginTop: Dimens.alertDividerMarginTop,
      height: Dimens.alertDividerHeight,
      backgroundColor: Colors.alertDividerColor,
    },
    actionContainer: {
      ...Layout.row,
      ...Layout.justifyContentSpaceEvenly,
      height: Dimens.alertActionHeight,
    },
    actionButton: {
      ...Layout.fill,
      borderRadius: Dimens.zero,
    },
    textStyle: {
      ...Fonts.textRegular,
      color: Colors.alertTextColor,
      padding: Dimens.alertActionTextPadding,
    },
    blurView: {
      ...Layout.fill,
    },
  })
}
