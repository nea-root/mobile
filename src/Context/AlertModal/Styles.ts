import { CommonParams } from '@/Theme/theme'
import { StyleSheet } from 'react-native'

export default function <C>({
  Colors,
  Layout,
  Gutters,
  Fonts,
  MetricsSizes,
  ViewSizes,
  Dimens,
}: CommonParams<C>) {
  return StyleSheet.create({
    blurView: {
      ...Layout.fill,
    },
  })
}
