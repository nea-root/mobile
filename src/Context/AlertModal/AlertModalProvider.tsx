import { ReactNode, createContext, useCallback, useMemo, useRef } from 'react'
import { AlertModalContext } from '../Types/AlertModalContext'
import { AlertModal } from '@/Components/CommonModal/AlertModal'
import { useModal, useTheme } from '@/Hooks'
import { AlertModalActionData } from '@/Data/Models/Application/AlertModalData'

import { BlurView } from '@react-native-community/blur'
import createStyles from './Styles'

export const AlertModalProvider = createContext<AlertModalContext>({
  alertModalData: undefined,
  showAlert: undefined,
  hideModal: undefined,
})

type Props = {
  children: ReactNode
}

export const AlertModalProviderContext = ({ children }: Props) => {
  const { alertModalData, showAlert, hideModal } = useModal()
  const isReady = useRef<boolean>(true)

  const theme = useTheme()
  const styles = createStyles(theme)

  let tmpShowAlert = useCallback(
    (
      isShow: boolean,
      title: string,
      description: string,
      actions: AlertModalActionData[] = [
        {
          label: 'OK',
          action: () => {
            hideModal()
            isReady.current = true
          },
        },
      ],
      isHighPriority: boolean = false,
    ) => {
      if (isHighPriority || isReady.current) {
        isReady.current = false
        showAlert(isShow, title, description, actions)
        return true
      }
      return false
    },
    [hideModal, showAlert],
  )

  let tmpHideModal = useCallback(() => {
    isReady.current = true
    hideModal()
  }, [hideModal])

  const modalProvider = useMemo(() => {
    return {
      alertModalData,
      showAlert: tmpShowAlert,
      hideModal: tmpHideModal,
    }
  }, [alertModalData, tmpHideModal, tmpShowAlert])

  return (
    <AlertModalProvider.Provider value={modalProvider}>
      {children}
      {
        alertModalData.isShow ?
          <BlurView
            autoUpdate={false}
            blurAmount={theme.Dimens.controlBackgroundBlurAmount}
            blurRadius={theme.Dimens.controlBackgroundBlurRadius}
            style={{ position: 'absolute', height: '100%', width: '100%' }}
          >
            <AlertModal {...alertModalData} />
          </BlurView>
          : <></>
      }
    </AlertModalProvider.Provider>
  )
}
