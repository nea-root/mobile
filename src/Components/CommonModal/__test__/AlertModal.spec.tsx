import { fireEvent, render, waitFor } from '@testing-library/react-native'
import { AlertModal } from '@/Components/CommonModal/AlertModal'
import { RenderResult } from '@testing-library/react-native/build/render'
import { AlertModalData } from '@/Data/Models/Application/AlertModalData'
import { Provider } from 'react-redux'
import { store } from '@/Store'

const props: AlertModalData = {
  isShow: true,
  title: 'AlertModel Title',
  description: 'AlertModal description',
  actions: [
    {
      label: 'Ok',
      testID: 'ok',
      action: jest.fn(),
    },
  ],
}
let renderResult: RenderResult
beforeEach(() => {
  renderResult = render(
    <Provider store={store}>
      <AlertModal {...props} />
    </Provider>,
  )
})

it('should render AlertModel correctly', async () => {
  await waitFor(() => {
    expect(renderResult.toJSON()).toMatchSnapshot()
  })
})

it('should onAccept function called when button is pressed', async () => {
  fireEvent.press(renderResult.getByTestId('alert_action_button_ok-text'))

  await waitFor(() => {
    if (!props.actions) {
      expect(props.actions).not.toBeNull()
      return
    }
    expect(props.actions[0].action).toBeCalled()
  })
})
