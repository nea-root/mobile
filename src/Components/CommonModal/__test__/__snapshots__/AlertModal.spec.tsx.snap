// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AlertModel correctly 1`] = `
<Modal
  hardwareAccelerated={false}
  testID="alertModal"
  transparent={true}
  visible={true}
>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "alignSelf": "stretch",
          "backgroundColor": "#ffffff",
          "borderRadius": undefined,
          "elevation": undefined,
          "marginHorizontal": 24,
          "paddingTop": 24,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": undefined,
            "width": undefined,
          },
          "shadowOpacity": undefined,
          "shadowRadius": undefined,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#25042D",
              "fontFamily": "NunitoSans-Bold",
              "fontSize": 20,
              "marginHorizontal": undefined,
              "textAlign": "center",
            },
          ]
        }
      >
        AlertModel Title
      </Text>
      <Text
        style={
          [
            {
              "color": "#25042D",
              "fontFamily": "NunitoSans-Regular",
              "fontSize": 16,
              "fontWeight": "600",
              "marginHorizontal": 24,
              "marginTop": undefined,
              "textAlign": "center",
            },
          ]
        }
      >
        AlertModal description
      </Text>
      <View
        style={
          [
            {
              "backgroundColor": "rgba(202, 196, 208, 1)",
              "height": 0.5,
            },
            undefined,
            false,
            false,
            [
              {
                "alignSelf": "stretch",
                "backgroundColor": "#A09BFF",
                "height": undefined,
                "marginTop": undefined,
              },
            ],
          ]
        }
      />
      <View
        style={
          {
            "flexDirection": "row",
            "height": undefined,
            "justifyContent": "space-evenly",
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderRadius": undefined,
              "flex": 1,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="alert_action_button_ok-container-outer-layer"
        >
          <View
            collapsable={false}
            removeClippedSubviews={false}
            style={
              {
                "backgroundColor": "transparent",
                "borderColor": "transparent",
                "borderRadius": undefined,
                "borderStyle": "solid",
                "borderWidth": 0,
                "flex": undefined,
                "minWidth": 64,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="alert_action_button_ok-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 20,
                  },
                ]
              }
              testID="alert_action_button_ok"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 12,
                        },
                        undefined,
                        false,
                        {
                          "color": "rgba(103, 80, 164, 1)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        {
                          "color": "#25042D",
                          "fontFamily": "NunitoSans-Regular",
                          "fontSize": 20,
                          "padding": undefined,
                        },
                      ],
                    ]
                  }
                  testID="alert_action_button_ok-text"
                >
                  Ok
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
